docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,556 @@
+
+2. Because we scan it too frequently, we'll spend a lot of CPU on building this queue, since we'll be scanning all the stats cache and the schema cache. See more: <https://github.com/pingcap/tidb/issues/49972>
+We can try to keep the priority queue in memory to solve the above issues instead of rebuilding it every time.
+
+## Current Implementation
+
+### Calculate the score
+
+In the current implementation, we use a weighted scoring approach to maintain the queue, which is our current scoring criteria.
+
+Weight table:
+
+| Name                 | Meaning                                                                                                                                                                                       | Weight                                                                                                                                                                                                                                                                                                              | Data Source                  |
+| -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------- |
+| Percentage of Change | The percentage of change since the last analysis.                                                                                                                                             | log10(1 + Change Ratio). [Check the graph](https://www.desmos.com/calculator/gb3i2cropz). <br><br> Note: For unanalyzed tables, we set the percentage of changes to 100%                                                                                                                                            | Stats Cache                  |
+| Table Size           | The size is equal to the number of rows multiplied by the number of columns in the table that are subject to analysis. The smaller tables should have a higher priority than the bigger ones. | Applying a logarithmic transformation, namely log10(1 + Table Size), and its 'penalty' is calculated as 1 - log10(1 + Table Size). [Check the graph.](https://www.desmos.com/calculator/x3fq76w8sb)                                                                                                                 | Stats Cache + Table Info     |
```

Comment:
maybe directly descibe the weight as "1 - log10(1 + Table Size)". 

BTW I guess the final score of a table is sum_for_factors(weight)? Then I think no need to use `1` in this formula, "-log10(1 + Table Size)" should be equivalent.

Reply:
> BTW I guess the final score of a table is sum_for_factors(weight)? Then I think no need to use `1` in this formula, "-log10(1 + Table Size)" should be equivalent.

If I understand correctly, they have the same slope, but you can look at their graphs, they have different value domains, the whole function is shifted up(down) on the image, so it will produce a different result for the formula, it will produce a different result for the same table size, and we want to use the first one, we don't want to penalize very much when the table size is very small.

Reply:
I have see the final formula `priority_score = ...` in your below comment. If every score minus `0.1` which means use -log10(...), the relative order is not changed.

Reply:
Yes, for DML changes, the relative order remains unchanged, but the value domain does change. I guess EventNewIndex will take a higher rank in some cases because it uses a fixed number score, which is 2.

Reply:
Discussion offline: We don't want to change the formula in this design. However, this change will affect the order between the newly added index job and other jobs. Therefore, we will need to adjust the value of the newly added index event as well.